### ISSUE TITLE
Cannot combine Lists, TypeVars, UnionTypes, and Callables in parameter typehints

### ISSUE THREAD

**Issue by pydsigner** [microsoft/pyright#3802]
If a function has two parameters, one a function returning values of type T, and the second a list of union of functions taking type T and None, no literal list can be constructed to pass in to the second argument that pyright will not reject.

```python
from __future__ import annotations

import typing as t


T = t.TypeVar('T')


def str2int(a: str) -> int:
    return int(a)

def int2str(b: int) -> str:
    return str(b)


def test(func_1: t.Callable[[str], T], varied: list[t.Callable[[T], str] | None]):
    pass

def test_nonone(func_1: t.Callable[[str], T], varied: list[t.Callable[[T], str]]):
    pass

def test_nocallable(func_1: t.Callable[[str], T], varied: list[T | None]):
    pass

def test_nogen(func_1: t.Callable[[str], int], varied: list[t.Callable[[int], str] | None]):
    pass

def test_sequence(func_1: t.Callable[[str], T], varied: t.Sequence[t.Callable[[T], str] | None]):
    pass


test(str2int, [int2str]) # fails!
test(str2int, [None]) # fails!
test(str2int, []) # fails!
test(str2int, [int2str, None]) # fails!
test_nonone(str2int, [int2str])
test_nonone(str2int, [])
test_nocallable(str2int, [5])
test_nocallable(str2int, [None])
test_nocallable(str2int, [5, None])
test_nogen(str2int, [int2str])
test_nogen(str2int, [None])
test_nogen(str2int, [])
test_nogen(str2int, [int2str, None])
test_sequence(str2int, [int2str])
test_sequence(str2int, [None])
test_sequence(str2int, [])
test_sequence(str2int, [int2str, None])
```

I expect that each test case will pass the checker. Instead, the arguments to `test` are inferred as `list[((int) -> str) | None]` and rejected:

```
Assuming Python platform Windows
Searching for source files
Found 1 source file
pyright 1.1.265
C:\Users\pydsigner\programming\anchovy\examples\min.py
  C:\Users\pydsigner\programming\anchovy\examples\min.py:32:15 - error: Argument of type "list[((int) -> str) | None]" cannot be assigned to parameter "varied" of type "list[((T@test) -> str) | None]" in function "test"
    TypeVar "_T@list" is invariant
      Type "((int) -> str) | None" cannot be assigned to type "((T@test) -> str) | None"
        Type "None" cannot be assigned to type "((T@test) -> str) | None"
          Type "None" cannot be assigned to type "(T@test) -> str" (reportGeneralTypeIssues)
  C:\Users\pydsigner\programming\anchovy\examples\min.py:33:15 - error: Argument of type "list[((int) -> str) | None]" cannot be assigned to parameter "varied" of type "list[((T@test) -> str) | None]" in function "test"
    TypeVar "_T@list" is invariant
      Type "((int) -> str) | None" cannot be assigned to type "((T@test) -> str) | None"
        Type "None" cannot be assigned to type "((T@test) -> str) | None"
          Type "None" cannot be assigned to type "(T@test) -> str" (reportGeneralTypeIssues)
  C:\Users\pydsigner\programming\anchovy\examples\min.py:34:15 - error: Argument of type "list[((int) -> str) | None]" cannot be assigned to parameter "varied" of type "list[((T@test) -> str) | None]" in function "test"
    TypeVar "_T@list" is invariant
      Type "((int) -> str) | None" cannot be assigned to type "((T@test) -> str) | None"
        Type "None" cannot be assigned to type "((T@test) -> str) | None"
          Type "None" cannot be assigned to type "(T@test) -> str" (reportGeneralTypeIssues)
  C:\Users\pydsigner\programming\anchovy\examples\min.py:35:15 - error: Argument of type "list[((int) -> str) | None]" cannot be assigned to parameter "varied" of type "list[((T@test) -> str) | None]" in function "test"
    TypeVar "_T@list" is invariant
      Type "((int) -> str) | None" cannot be assigned to type "((T@test) -> str) | None"
        Type "None" cannot be assigned to type "((T@test) -> str) | None"
          Type "None" cannot be assigned to type "(T@test) -> str" (reportGeneralTypeIssues)
4 errors, 0 warnings, 0 informations
Completed in 0.914sec
```

Note that switching from `list` to `typing.Sequence` solves the problem, perhaps related in root cause to #2659? However, I would not have expected that getting rid of `typing.Callable` would avoid the error as well.

I am running pyright 1.1.265 from the console as well as using Pylance v2022.8.11 in VS Code.

For completeness, I ran mypy 0.971 as well and received no complaints about `test()`, but curiously got an error about `test_nocallable(str2int, [None])`.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by pydsigner** [comment]
Confirmed that the issue is resolved, thanks @erictraut !
